File: LeetCode/Medium/1297-maximum-number-of-occurrences-of-a-substring/1297-maximum-number-of-occurrences-of-a-substring.js
/**
 * @param {string} s
 * @param {number} maxLetters
 * @param {number} minSize
 * @param {number} maxSize
 * @return {number}
 */
var maxFreq = function(s, maxLetters, minSize, maxSize) {
    let hashmap = {};
    let right = 0;
    let set = new Set();
    let curr = "";
    let ans = 0;

    for (let i = 0; i < s.length - minSize+1; i++) {
        right = i + minSize;
        curr = s.slice(i, right);
        set = new Set(curr);
        if (set.size > maxLetters) {
            continue;
        }
        hashmap[curr] = hashmap[curr] || 0;
        hashmap[curr] += 1;
    }

    for (let i in hashmap) {
        ans = hashmap[i] > ans ? hashmap[i] : ans;
    }

    return ans;
};

File: LeetCode/Medium/1297-maximum-number-of-occurrences-of-a-substring/1297-maximum-number-of-occurrences-of-a-substring.py
class Solution:
    def maxFreq(self, s: str, maxLetters: int, minSize: int, maxSize: int) -> int:
        hashmap, left = defaultdict(int), 0

        for right in range(len(s)):
            sets = set(s[left:right+1])

            if right - left + 1 == minSize:
                hashmap[s[left:right+1]] += 1 if len(sets) <= maxLetters else 0
                left += 1

        return max(hashmap.values()) if hashmap else 0

"""
-If a subarray of len maxSize is repeated, then all subarrays of that subarray were also repeated, so you only need to take note of minSize
-Use a sliding window of len minSize to track subarrays
-For each subarray, convert it to a set. If the length of the set is less than or equal to maxLetters, increase the value of the subarray in the hashmap by 1
-Return Max Hahsmap
"""